### PR TITLE
C++ Library for Decoding CAN messages from PE3 ECU

### PR DESCRIPTION
**Pull request by ethanmpeterson**
A sample PR for the git tutorial.

### CODE REVIEW

**Comment by bchampp** [comment]
Had some trouble fixing the Makefile for the CI pipeline. Please run 'make' before pushing any code review changes to ensure it'll pass the pipeline

File: CAN/PE3/ECU.h [hunk 1]
```diff
@@ -0,0 +1,31 @@
+#ifndef ECU_h
+#define ECU_h
+
+#include "Arduino.h"
```
Comment (bchampp) [inline CAN/PE3/ECU.h:4]: Where is this file located?

File: CAN/PE3/ECU.h [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  bool pressureUnit;
+  float MAP;
+  float lambda;
+  float analogInputs[8];
+  float frequencies[4];
+  float batVoltage;
+  float airTemp;
+  float coolantTemp;
+  bool tempUnit;
+  float thermistors[2];
```
Comment (bchampp) [inline CAN/PE3/ECU.h:25]: Why is this an array?

File: CAN/PE3/ECU.h [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+public:
+  ECU();
+  void update(unsigned long id, unsigned char msgBuf[], unsigned char msgLen);
+  void debugPrint(unsigned long id);
+  int rpm;
+  float tps;
+  float openTime;
+  float ignitionAngle;
+  float pressure;
+  bool pressureUnit;
```
Comment (bchampp) [inline CAN/PE3/ECU.h:16]: What are the options ? Not specified.

File: CAN/PE3/ECU.cpp [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    for (int i = 0; i < 4; i += 2) {
+      this->thermistors[i / 2] = readSigned(msgBuf[i], msgBuf[i + 1], 0.1);
+    }
+    this->versionMajor = readUnsigned(msgBuf[4], 0, 1);
+    this->versionMinor = readUnsigned(msgBuf[5], 0, 1);
+    this->versionBuild = readUnsigned(msgBuf[6], 0, 1);
+    break;
+  }
+}
+
```
Comment (bchampp) [inline CAN/PE3/ECU.cpp:83]: There is a better way to do this I think

File: CAN/PE3/DEFS.h [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+#define PE7 0x0000F648
+#define PE8 0x0000F748
+#define PE9 0x0000F848
+#define PE10 0x0000F948
+
+#define PE11 0x0000FA48
+#define PE12 0x0000FB48
+#define PE13 0x0000FC48
+#define PE14 0x0000FD48
+#define PE15 0x0000FE48
```
Comment (bchampp) [inline CAN/PE3/DEFS.h:30]:
```
`
ECU extends CAN {
// Put specific addresses here

}
```
`

File: CAN/PE3/QCAN.ino [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+  NOTE: Currently only the PE1 - PE7 addresses are supported by the ECU library
+*/
+
+#include "DEFS.h"
+#include "ECU.h"
+#include "mcp_can.h"
+
+// Use pin 10 for CAN shield version 1.2
+#define SPI_CS_PIN 10
```
Comment (bchampp) [inline CAN/PE3/QCAN.ino:16]:
```suggestion
#define SPI_CS_PIN 11
```

**Review by bchampp — CHANGES_REQUESTED** [review]
Abstract into a CAN.h and CAN.cpp for varying addresses.

File: CAN/PE3/ECU.h (hunk 3 above)
Comment (ethanmpeterson) [reply CAN/PE3/ECU.h:16]: I will change this to a tuple or something more descriptive.

**Review by bchampp — APPROVED** [review]

**Pull request merged**